extendedcopy.go

```diff
@@ -196,3 +199,75 @@ func findRoots(ctx context.Context, storage content.GraphStorage, node ocispec.D
 	}
 	return roots, nil
 }
+
+// FilterArtifactType will configure opts.FindPredecessors to filter the predecessors
+// whose artifact type matches a given regex pattern.
+func (opts *ExtendedCopyGraphOptions) FilterArtifactType(regex *regexp.Regexp) {
+	fp := opts.FindPredecessors
+	opts.FindPredecessors = func(ctx context.Context, src content.GraphStorage, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		var predecessors []ocispec.Descriptor
+		var err error
+		if fp == nil {
+			// if src is a ReferrerFinder, use Referrers() to filter the predecessors.
+			rf, ok := src.(registry.ReferrerFinder)
+			if ok {
+				predecessors, err = findReferrersAndFilter(rf, ctx, desc, regex)
+				if err != nil {
+					return predecessors, nil
+				}
+			} else {
+				predecessors, err = src.Predecessors(ctx, desc)
+			}
```

Comment:
The error handling is incorrect.
```suggestion
			if rf, ok := src.(registry.ReferrerFinder); ok {
				return findReferrersAndFilter(rf, ctx, desc, regex)
			}
			predecessors, err = src.Predecessors(ctx, desc)
```

Reply:
Changed accordingly.